src/diff.jl

```diff
@@ -0,0 +1,228 @@
+                                ),
+                            )
+                        end
+                    end
+                    MOI.set(
+                        model.optimizer,
+                        DiffOpt.ForwardConstraintFunction(),
+                        inner_ci,
+                        MOI.ScalarAffineFunction{T}(terms, cte),
+                    )
+                end
+            end
+        end
+    end
+    if model.affine_objective_cache !== nothing
+        cte = zero(T)
+        terms = MOI.ScalarAffineTerm{T}[]
+        pf = model.affine_objective_cache
+        sizehint!(terms, 0)
+        if length(pf.p) != 0
+            for term in pf.p
+                p = p_idx(term.variable)
+                sensitivity = get(model.parameter_input_forward, p, 0.0)
+                # TODO: check sign
+                cte += sensitivity * term.coefficient
+            end
+            # TODO: if cte != 0
+            MOI.set(
+                model.optimizer,
+                DiffOpt.ForwardObjectiveFunction(),
+                inner_ci,
+                MOI.ScalarAffineFunction{T}(terms, cte),
+            )
+        end
+    elseif model.quadratic_objective_cache !== nothing
+        cte = zero(T)
+        terms = MOI.ScalarAffineTerm{T}[]
+        pf = model.quadratic_objective_cache
+        sizehint!(terms, length(pf.pv))
+        if length(pf.p) != 0
+            for term in pf.p
+                p = p_idx(term.variable)
+                sensitivity = get(model.parameter_input_forward, p, 0.0)
+                # TODO: check sign
+                cte += sensitivity * term.coefficient
+            end
+            for term in pf.pp
+                p_1 = p_idx(term.variable_1)
+                p_2 = p_idx(term.variable_2)
+                sensitivity_1 = get(model.parameter_input_forward, p_1, 0.0)
+                sensitivity_2 = get(model.parameter_input_forward, p_2, 0.0)
+                cte += sensitivity_1 * sensitivity_2 * term.coefficient
+            end
+            # canonicalize?
+            for term in pf.pv
+                p = p_idx(term.variable_1)
+                sensitivity = get(model.parameter_input_forward, p, NaN)
+                if !isnan(sensitivity)
+                    push!(
+                        terms,
+                        MOI.ScalarAffineTerm{T}(
+                            sensitivity * term.coefficient,
+                            term.variable_2,
+                        ),
+                    )
+                end
+            end
+            # TODO: if cte != 0
+            MOI.set(
+                model.optimizer,
+                DiffOpt.ForwardObjectiveFunction(),
+                inner_ci,
+                MOI.ScalarAffineFunction{T}(terms, cte),
+            )
+        end
+    end
+    DiffOpt.forward_differentiate!(model.optimizer)
+    return
+end
+
+struct ForwardParameter <: MOI.AbstractVariableAttribute end
```

Comment:
We can consider that. On the other hand, ForwardVariablePrimal is for output sensitivity, while ForwardParameter is for input sensitivity. Having both different would be good for validation, since we gave up on names like: ForwardOutputVariablePrimal.

Reply:
You can different between input and output depending on whether it's a set or a get. Note that defining a new struct or a new function isn't so natural for an extension, it's more designed to add methods for existing ones. However, it is possible, see the hack in NLopt.

Reply:
Adding more details to my previous comment:
I find it strange that:
For parameters:
`set` `ForwardVariablePrimal` sets an input value that can be `get` to check which value was there.
While, for actual variables:
`set` `ForwardVariablePrimal` always errors, and `get` `ForwardVariablePrimal` only makes sense after `forward_differentiate!`
This was the main motivation for the new attribute.

About:

>Note that defining a new struct or a new function isn't so natural for an extension, it's more designed to add methods for existing ones.

I think I did not understand completely. DiffOpt adds new structs. Also, a few solvers define new structs (like `Gurobi.NumberOfObjectives`, `GLPK.CallbackFunction`, `COSMO.ADMMIterations`).

Reply:
Ah! now I get it, you mean Julia extensions like `NLoptMathOptInterfaceExt.jl`

Reply:
Yes, I meant Julia extensions :) The workaround in NLopt works quite well but it becomes tricky when you try to include their docstring in the documentation. It was quite hard to make it work in https://github.com/JuliaManifolds/Manopt.jl/pull/264 for instance.
Also, as from the MOI level, these are also variables and the set can be bridged to `EqualTo`, it makes sense to consider it as `ForwardVariablePrimal`.

Actually, what we could do it add `ForwardConstraintSet` that is defined for `MOI.Parameter`, `MOI.EqualTo`, `MOI.LessThan`, `MOI.GreaterThan`, `MOI.Interval`. I think we worked around it in DiffOpt by using the constant in the function but if you have a `VariableIndex`-in-`S` then you can't modify the function right ?
We could disallow `ForwardConstraintSet` for non-`VariableIndex` to avoid having two ways to set the same thing. Even if that's not consistent with the `ConstraintFunction`/`ConstraintSet` attributes, that's backward compatible.